assets/js/data/cart/actions.js

```diff
@@ -0,0 +1,149 @@
+/**
+ * External dependencies
+ */
+import { apiFetch, select } from '@wordpress/data-controls';
+
+/**
+ * Internal dependencies
+ */
+import { ACTION_TYPES as types } from './action-types';
+import { MISSING_ROUTE_ERROR } from './constants';
+import { STORE_KEY as SCHEMA_STORE_KEY } from '../schema/constants';
+
+/**
+ * Returns an action object used in updating the store with the provided items
+ * retrieved from a request using the given querystring.
+ *
+ * This is a generic response action.
+ *
+ * @param {Object}   [response={}]    An object containing the response from the
+ *                                    request.
+ * @return {Object} Object for action.
+ */
+export function receiveCart( response = {} ) {
+	return {
+		type: types.RECEIVE_CART,
+		response,
+	};
+}
+
+/**
+ * Returns an action object used for receiving customer facing errors from the
+ * API.
+ *
+ * @param {Object}  [error={}]     An error object containing the error message
+ *                                 and response code.
+ * @param {boolean} [replace=true] Should existing errors be replaced, or should
+ *                                 the error be appended.
+ * @return {Object} Object for action.
+ */
+export function receiveError( error = {}, replace = true ) {
+	return {
+		type: replace ? types.REPLACE_ERRORS : types.RECEIVE_ERROR,
+		error,
+	};
+}
+
+/**
+ * Returns an action object used to track when a coupon is applying.
+ *
+ * @param {boolean} [couponCode] Coupon being added.
+ * @return {Object} Object for action.
+ */
+export function receiveApplyingCoupon( couponCode ) {
+	return {
+		type: types.APPLYING_COUPON,
+		couponCode,
+	};
+}
+
+/**
+ * Returns an action object used to track when a coupon is removing.
+ *
+ * @param {string} [couponCode] Coupon being removed.
+ * @return {Object} Object for action.
+ */
+export function receiveRemovingCoupon( couponCode ) {
+	return {
+		type: types.REMOVING_COUPON,
+		couponCode,
+	};
+}
+
+/**
+ * Applies a coupon code and either invalidates caches, or receives an error if
+ * the coupon cannot be applied.
+ *
+ * @param {string} couponCode The coupon code to apply to the cart.
+ */
+export function* applyCoupon( couponCode ) {
+	yield receiveApplyingCoupon( couponCode );
+
+	try {
+		const route = yield select(
+			SCHEMA_STORE_KEY,
+			'getRoute',
+			'/wc/store',
+			'cart/apply-coupon/(?P<code>[\\w-]+)',
+			[ couponCode ]
+		);
+
+		if ( ! route ) {
+			yield receiveError( MISSING_ROUTE_ERROR );
+			return;
+		}
```

Comment:
Since `applyCoupon`(applies for other action creators in here as well) is a specific type of action here, I think it'd be okay to just pass the known route directly to `apiFetch` and skip going through the schema store?